src/lib/getPriorityIndex.ts

```diff
@@ -299,29 +351,112 @@ const priorityGroups: readonly PriorityGroup[] = [
+  // getPriority returns a number. The smaller is the heigher priority.
+  // Properties will have their "group priority", determined by which group property belongs, and "inGroup priority", determined by index in that group.
+  const { firstProps, lastProps, componentSpecificProps } = config;
+  const indexInFirstProps = firstProps.indexOf(key);
+  const indexInLastProps = lastProps.indexOf(key);
+
+  if (indexInFirstProps !== -1) {
+    return calcPriorityFromIndex({ type: "reservedFirstProps", value: indexInFirstProps }, config);
+  }
+  if (indexInLastProps !== -1) {
+    return calcPriorityFromIndex({ type: "reservedLastProps", value: indexInLastProps }, config);
+  }
+
+  if (componentSpecificProps) {
+    const index = componentSpecificProps.indexOf(key);
+    if (index !== -1) {
+      return calcPriorityFromIndex({ type: "componentSpecificProps", value: index }, config);
+    }
+  }
+
+  // Then it can be either `stylePropsPriority` or `otherPropsPriority`
+  const groupIndex = priorityGroups.findIndex((group) => {
     return group.keys.includes(key);
   });
-  return index === -1 ? Number.MAX_SAFE_INTEGER : index;
+
+  const isStyleProps = groupIndex > -1;
+  if (isStyleProps) {
+    const keyIndex = getIndexInGroup(key, groupIndex);
+    return calcPriorityFromIndex({ type: "styleProps", groupIndex, keyIndex }, config);
+  }
+
+  return calcPriorityFromIndex({ type: "otherProps" }, config);
 }
 
-export const priorityGroupsLength = priorityGroups.length;
+type Index =
+  | { type: "reservedFirstProps" | "componentSpecificProps" | "reservedLastProps"; value: number }
+  | { type: "styleProps"; groupIndex: number; keyIndex: number }
+  | { type: "otherProps" };
 
-export const getIndexInPriority = (key: string, groupIndex: number): number => {
-  const keys = priorityGroups[groupIndex].keys;
-  if (keys) {
-    const index = keys.indexOf(key);
-    return index === -1 ? Number.MAX_SAFE_INTEGER : index;
+const calcPriorityFromIndex = (index: Index, config: Config) => {
+  // Calculates the priority in which every property has different priority.
+  // As an exception, non-predefined properties have the same precedence.
+  // They will be treated as "other Props".
+
+  // Currently, the priority is determined from the index of the array.
+  // We assume that the length of each array is atmost 100.
+  // When changing the specification, be sure to check that the stylePropsPriority range does not overlap with others.
+  // Now it's range is [200, 10200]. 10200 is 100 * 100 + 200.
+
+  // Perhaps we may want to handle -1 as error in some future.
+  // Therefore I set the priority to numbers greater than or equal to zero.
+  const isComponentSpecBeforeStyle = config.isCompPropsBeforeStyleProps;
+  const basePriorities = {
+    firstProps: 0,
+    styleProps: 200,
+    componentSpecificProps: isComponentSpecBeforeStyle ? 100 : 10 * 5,
+    otherProps: 10 ** 6,
+    lastProps: 10 ** 6 + 1,
+  };
```

Comment:
I think the priority ranges are little complex. How about this?
- The ten thousands place value is always represent `basePriorities`.
- The range of `stylePropsPriority` is 0~99 so the max priority of `styleProps` is `29999`.
```suggestion
  const basePriorities = {
    firstProps: 0,
    styleProps: 20000,
    componentSpecificProps: isComponentSpecBeforeStyle ? 10000 : 30000,
    otherProps: 40000,
    lastProps: 50000,
  };
```

Reply:
Looks great!